### PR TITLE
rgw_sal_motr: [CORTX-28987] Fix object deletion with null version

### PR DESCRIPTION
**Pull request by SeagateChaDeepak**
# Change Summary

**Problem :** If version id parameter is passed as null to delete-object
API in case of unversioned bucket for deletion of simple object,
then object deletion does not happen.

**Implementation :** If instance is set to "null" we clear this string
in delete_object implementation. This implementation is similar to
what is done in rados.

## Parent Story
_[CORTX-28987](https://jts.seagate.com/browse/CORTX-28987):_
_Deletion of simple object with null version id._

## Sub Tasks covered in this PR
NA

## Exclusions in Implementation :
Multipart object not tested.

## Testing Done : 

<details>
  <summary> **Case 1 : Deleting Simple Object in unversioned bucket** </summary>
  

> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 mb s3://unversionedtestbucket --endpoint-url http://127.0.0.1:8000
> make_bucket: unversionedtestbucket
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 cp 1mfile s3://unversionedtestbucket --endpoint-url http://127.0.0.1:8000
> upload: ./1mfile to s3://unversionedtestbucket/1mfile
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api list-object-versions --bucket unversionedtestbucket --endpoint-url http://127.0.0.1:8000
> {
>     "Versions": [
>         {
>             "ETag": "\"b6d81b360a5672d80c27430f39153e2c\"",
>             "Size": 1048576,
>             "StorageClass": "STANDARD",
>             "Key": "1mfile",
>             "VersionId": "null",
>             "IsLatest": true,
>             "LastModified": "2022-02-28T07:22:50.630Z",
>             "Owner": {
>                 "DisplayName": "mgwuser1",
>                 "ID": "mgwuser1"
>             }
>         }
>     ]
> }
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api delete-object --bucket unversionedtestbucket --key 1mfile --endpoint-url http://127.0.0.1:8000 --version-id null
> 
> An error occurred (InvalidArgument) when calling the DeleteObject operation: Unknown
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api delete-object --bucket unversionedtestbucket --key 1mfile --endpoint-url http://127.0.0.1:8000 --version-id null
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 ls s3://unversionedtestbucket --endpoint-url http://127.0.0.1:8000
> [root@ssc-vm-g4-rhev4-0344 build]#

</details>

<details>
  <summary> **Case 2 : Deleting Simple Object in versioned bucket** </summary>

> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 mb s3://versionedtestbucket --endpoint-url http://127.0.0.1:8000
> make_bucket: versionedtestbucket
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api put-bucket-versioning --bucket versionedtestbucket --versioning-configuration Status=Enabled --endpoint-url http://127.0.0.1:8000
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api get-bucket-versioning --bucket versionedtestbucket --endpoint-url http://127.0.0.1:8000
> {
>     "Status": "Enabled",
>     "MFADelete": "Disabled"
> }
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 cp 1mfile s3://versionedtestbucket --endpoint-url http://127.0.0.1:8000
> upload: ./1mfile to s3://versionedtestbucket/1mfile
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api list-object-versions --bucket versionedtestbucket --endpoint-url http://127.0.0.1:8000
> {
>     "Versions": [
>         {
>             "ETag": "\"b6d81b360a5672d80c27430f39153e2c\"",
>             "Size": 1048576,
>             "StorageClass": "STANDARD",
>             "Key": "1mfile",
>             "VersionId": "uKMNszkMI9VYw0gLhGzMoeVkCu.y7-F",
>             "IsLatest": true,
>             "LastModified": "2022-02-28T07:33:09.817Z",
>             "Owner": {
>                 "DisplayName": "mgwuser1",
>                 "ID": "mgwuser1"
>             }
>         }
>     ]
> }
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api delete-object --bucket versionedtestbucket --key 1mfile --endpoint-url http://127.0.0.1:8000 --version-id uKMNszkMI9VYw0gLhGzMoeVkCu.y7-F
> 
> An error occurred (InvalidArgument) when calling the DeleteObject operation: Unknown
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3api delete-object --bucket versionedtestbucket --key 1mfile --endpoint-url http://127.0.0.1:8000 --version-id uKMNszkMI9VYw0gLhGzMoeVkCu.y7-F
> [root@ssc-vm-g4-rhev4-0344 build]# aws s3 ls s3://versionedtestbucket --endpoint-url http://127.0.0.1:8000
> [root@ssc-vm-g4-rhev4-0344 build]#

</details>

Signed-off-by: Deepak Mahale <deepak.mahale@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by sachinpunadikar — APPROVED** [review]
+1

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -1523,6 +1523,9 @@ struct rgw_obj_key {
     if (instance.empty()) {
       return name;
     }
+    if(have_null_instance()){
```
Comment (bkirunge7) [inline src/rgw/rgw_common.h:1526]: Please verify changes with rgw+rados

Comment (SeagateChaDeepak) [reply src/rgw/rgw_common.h]: Changed implementation as per discussion.

File: src/rgw/rgw_common.h [hunk 2]
```diff
@@ -1527,6 +1527,17 @@ struct rgw_obj_key {
     snprintf(buf, sizeof(buf), "%s[%s]", name.c_str(), instance.c_str());
     return buf;
   }
+
+  // Helper function to find the key of an object.
+  // in case of unversioned bucket object key is returned as is.
+  // in case of versioned bucket object key is key[version_id].
```
Comment (sjain09) [inline src/rgw/rgw_common.h:1533]: Mention in the comment that this additional function is for handling the null version if specified in the request.

Comment (SeagateChaDeepak) [reply src/rgw/rgw_common.h:1534]: Done

**Review by sjain09 — APPROVED** [review]
Looks good

File: src/rgw/rgw_common.h [hunk 3]
```diff
@@ -1527,6 +1527,18 @@ struct rgw_obj_key {
...
+  // This function is for handling the null version if specified in the request.
+  // in case of unversioned bucket object key is returned as is.
+  // in case of versioned bucket object key is key[version_id].
+  std::string get_key_name_string() const{
+    if (have_null_instance()) {
+      return name;
+    }
+
+    return to_str();
+  }
```
Comment (andriytk) [inline src/rgw/rgw_common.h:1541]: This should be reviewed at ceph repo by rgw devs. Please, create a PR against ceph repo and suggest this change there.

Comment (DPG17) [reply src/rgw/rgw_common.h:1541]: @SeagateChaDeepak : We should clear the instance when the user specifies the null version id in Del obj API. This is already done in Rados code, see in: RGWRados::Object::Delete::delete_obj().
 _if (instance == "null") {
    obj.key.instance.clear();
  }_

Similarly, in our Motr sal delete_obj code, we should reset instance when version id is "null".
This won't require any change in common code.

Comment (DPG17) [reply src/rgw/rgw_common.h:1541]: @SeagateChaDeepak : Your previous code change (get_object_key_str()) in Motr sal layer looks more feasible then this common code change.

**Comment by andriytk** [comment]
Please, update the subject line (make it in imperative mood) and elaborate the PR description. It should explain: 1) what the problem is (from which it should be obvious why we need this fix); 2) how we are fixing it.

**Comment by andriytk** [comment]
I like the description you did at https://github.com/ceph/ceph/pull/45292. Why not having the same explanation here?

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -1500,6 +1500,14 @@ MotrObject::MotrDeleteOp::MotrDeleteOp(MotrObject *_source, RGWObjectCtx *_rctx)
 // 2. Delete an object when its versioning is turned on.
 int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional_yield y)
 {
+  if (source->have_instance()) {
+    rgw_obj_key& key = source->get_key();
+    const std::string& instance = key.get_instance();
+    if (instance == "null") {
+      key.set_instance("");
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:1507]: Why not `key.instance.clear()` ?

Also, style: if there is only one line under `if` - there is no need for braces:

```
if (instance == "null")
  key.instance.clear()
```

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -1508,6 +1516,7 @@ int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional
     return rc;
   }
 
+
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:1519]: Is this really needed? If not - please remove.

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done the change.
I thought that using getters and setters would be better than accessing the internal member of structure.

File: src/rgw/rgw_sal_motr.cc (hunk 5 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Removed

**Review by andriytk — APPROVED** [review]

**Comment by andriytk** [comment]
Please fix the subject line of the commit msg also.

**Review by bkirunge7 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -1500,6 +1500,13 @@ MotrObject::MotrDeleteOp::MotrDeleteOp(MotrObject *_source, RGWObjectCtx *_rctx)
 // 2. Delete an object when its versioning is turned on.
 int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional_yield y)
 {
+  if (source->have_instance()) {
+    rgw_obj_key& key = source->get_key();
+    const std::string& instance = key.get_instance();
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

**Comment by andriytk** [comment]
One last minor thing - remove period at the end of the commit subject line. :)
https://cbea.ms/git-commit/#end

**Review by DPG17 — APPROVED** [review]

**Comment by andriytk** [comment]
> This case is valid only when Motr is used as backend for rgw.

I'm afraid, this sentence in the description is misleading. Isn't the case also valid in case of rados backend? Please, rephrase it.

**Pull request merged**